File: 13-15/JAVA/src/registry/AnimalRegistry.java
package registry;

import animals.*;
import animals.domestic.*;
import animals.pack.*;
import util.Counter;

import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

public class AnimalRegistry {
    private List<Animal> animals;
    private Counter counter;

    public AnimalRegistry() {
        this.animals = new ArrayList<>();
        this.counter = new Counter();
    }

    public void addAnimal(String animalType, String name, String birthDate) throws Exception {
        try (Counter c = counter) {
            Animal animal;
            switch (animalType.toLowerCase()) {
                case "dog":
                    animal = new Dog(name, birthDate);
                    break;
                case "cat":
                    animal = new Cat(name, birthDate);
                    break;
                case "hamster":
                    animal = new Hamster(name, birthDate);
                    break;
                case "horse":
                    animal = new Horse(name, birthDate);
                    break;
                case "camel":
                    animal = new Camel(name, birthDate);
                    break;
                case "donkey":
                    animal = new Donkey(name, birthDate);
                    break;
                default:
                    throw new IllegalArgumentException("Неизвестный тип животного");
            }
            animals.add(animal);
            System.out.printf("Животное %s добавлено с ID %d\n", name, animals.size());
        }
    }

    public void teachCommand(int animalId, String command) {
        if (animalId < 1 || animalId > animals.size()) {
            throw new IllegalArgumentException("Неверный ID животного");
        }
        animals.get(animalId - 1).addCommand(command);
        System.out.println("Команда добавлена");
    }

    public void listCommands(int animalId) {
        if (animalId < 1 || animalId > animals.size()) {
            throw new IllegalArgumentException("Неверный ID животного");
        }
        List<String> commands = animals.get(animalId - 1).getCommands();
        System.out.println("Команды животного: " + String.join(", ", commands));
    }

    public void showMenu() {
        Scanner scanner = new Scanner(System.in);
        while (true) {
            System.out.println("\nРеестр домашних животных");
            System.out.println("1. Завести новое животное");
            System.out.println("2. Обучить животное команде");
            System.out.println("3. Просмотреть команды животного");
            System.out.println("4. Выход");
            System.out.print("Выберите действие: ");

            String choice = scanner.nextLine();

            try {
                switch (choice) {
                    case "1":
                        System.out.print("Введите тип животного (dog, cat, hamster, horse, camel, donkey): ");
                        String type = scanner.nextLine();
                        System.out.print("Введите имя животного: ");
                        String name = scanner.nextLine();
                        System.out.print("Введите дату рождения (ГГГГ-ММ-ДД): ");
                        String birthDate = scanner.nextLine();
                        addAnimal(type, name, birthDate);
                        break;
                    case "2":
                        System.out.print("Введите ID животного: ");
                        int idToTeach = Integer.parseInt(scanner.nextLine());
                        System.out.print("Введите команду: ");
                        String command = scanner.nextLine();
                        teachCommand(idToTeach, command);
                        break;
                    case "3":
                        System.out.print("Введите ID животного: ");
                        int idToList = Integer.parseInt(scanner.nextLine());
                        listCommands(idToList);
                        break;
                    case "4":
                        scanner.close();
                        return;
                    default:
                        System.out.println("Неверный выбор");
                }
            } catch (Exception e) {
                System.out.println("Ошибка: " + e.getMessage());
            }
        }
    }
}

File: 13-15/C++/src/AnimalRegistry.cpp
#include "AnimalRegistry.h"
#include "Dog.h"
#include "Cat.h"
#include "Hamster.h"
#include "Horse.h"
#include "Camel.h"
#include "Donkey.h"
#include <iostream>
#include <stdexcept>
#include <algorithm>

void AnimalRegistry::addAnimal(const std::string &animalType, const std::string &name, const std::string &birthDate)
{
    try
    {
        Counter c;
        int id = c.add();

        std::unique_ptr<Animal> animal;
        if (animalType == "dog")
        {
            animal = std::make_unique<Dog>(name, birthDate);
        }
        else if (animalType == "cat")
        {
            animal = std::make_unique<Cat>(name, birthDate);
        }
        else if (animalType == "hamster")
        {
            animal = std::make_unique<Hamster>(name, birthDate);
        }
        else if (animalType == "horse")
        {
            animal = std::make_unique<Horse>(name, birthDate);
        }
        else if (animalType == "camel")
        {
            animal = std::make_unique<Camel>(name, birthDate);
        }
        else if (animalType == "donkey")
        {
            animal = std::make_unique<Donkey>(name, birthDate);
        }
        else
        {
            throw std::invalid_argument("Неизвестный тип животного");
        }

        animals.push_back(std::move(animal));
        std::cout << "Животное " << name << " добавлено с ID " << animals.size() << std::endl;

        c.close();
    }
    catch (const std::exception &e)
    {
        std::cerr << "Ошибка: " << e.what() << std::endl;
    }
}

void AnimalRegistry::teachCommand(int animalId, const std::string &command)
{
    if (animalId < 1 || animalId > static_cast<int>(animals.size()))
    {
        throw std::out_of_range("Неверный ID животного");
    }
    animals[animalId - 1]->addCommand(command);
    std::cout << "Команда добавлена" << std::endl;
}

void AnimalRegistry::listCommands(int animalId) const
{
    if (animalId < 1 || animalId > static_cast<int>(animals.size()))
    {
        throw std::out_of_range("Неверный ID животного");
    }
    auto commands = animals[animalId - 1]->getCommands();
    std::cout << "Команды животного: ";
    for (size_t i = 0; i < commands.size(); ++i)
    {
        if (i != 0)
            std::cout << ", ";
        std::cout << commands[i];
    }
    std::cout << std::endl;
}

void AnimalRegistry::showMenu()
{
    while (true)
    {
        std::cout << "\nРеестр домашних животных\n";
        std::cout << "1. Завести новое животное\n";
        std::cout << "2. Обучить животное команде\n";
        std::cout << "3. Просмотреть команды животного\n";
        std::cout << "4. Выход\n";
        std::cout << "Выберите действие: ";

        std::string choice;
        std::getline(std::cin, choice);

        try
        {
            if (choice == "1")
            {
                std::cout << "Введите тип животного (dog, cat, hamster, horse, camel, donkey): ";
                std::string type;
                std::getline(std::cin, type);

                std::cout << "Введите имя животного: ";
                std::string name;
                std::getline(std::cin, name);

                std::cout << "Введите дату рождения (ГГГГ-ММ-ДД): ";
                std::string birthDate;
                std::getline(std::cin, birthDate);

                addAnimal(type, name, birthDate);
            }
            else if (choice == "2")
            {
                std::cout << "Введите ID животного: ";
                int id;
                std::cin >> id;
                std::cin.ignore();

                std::cout << "Введите команду: ";
                std::string command;
                std::getline(std::cin, command);

                teachCommand(id, command);
            }
            else if (choice == "3")
            {
                std::cout << "Введите ID животного: ";
                int id;
                std::cin >> id;
                std::cin.ignore();

                listCommands(id);
            }
            else if (choice == "4")
            {
                break;
            }
            else
            {
                std::cout << "Неверный выбор\n";
            }
        }
        catch (const std::exception &e)
        {
            std::cerr << "Ошибка: " << e.what() << std::endl;
        }
    }
}
